spring-cloud-deployer-spi/src/main/java/org/springframework/cloud/deployer/spi/app/AbstractActuatorTemplate.java

```diff
@@ -0,0 +1,104 @@
+import org.springframework.web.util.UriComponentsBuilder;
+
+public abstract class AbstractActuatorTemplate implements ActuatorOperations {
+
+	protected final Log logger = LogFactory.getLog(getClass().getName());
+
+	protected final RestTemplate restTemplate;
+	protected final AppDeployer appDeployer;
+
+	protected AbstractActuatorTemplate(RestTemplate restTemplate, AppDeployer appDeployer) {
+		this.restTemplate = restTemplate;
+		this.appDeployer = appDeployer;
+	}
+
+	@Override
+	public <T> T getFromActuator(String deploymentId, String guid, String endpoint, Class<T> responseType) {
+		String actuatorUrl = getActuatorUrl(deploymentId, guid);
+
+		ResponseEntity<T> responseEntity = get(UriComponentsBuilder
+				.fromHttpUrl(actuatorUrl).path(normalizePath(endpoint)).toUriString(), responseType);
+		if (responseEntity.getStatusCode().isError()) {
+			logger.error(responseEntity.getStatusCode().toString());
+		}
+		return responseEntity.getBody();
+	}
+
+	@Override
+	public <T, R> R postToActuator(String deploymentId, String guid, String endpoint, T body,
+			Class<R> responseType) {
+		String actuatorUrl = getActuatorUrl(deploymentId, guid);
+		ResponseEntity<R> responseEntity = post(UriComponentsBuilder
+				.fromHttpUrl(actuatorUrl).path(normalizePath(endpoint)).toUriString(), body, responseType);
+		if (responseEntity.getStatusCode().isError()) {
+			logger.error(responseEntity.getStatusCode().toString());
+		}
+		return responseEntity.getBody();
+	}
+
+	protected HttpHeaders requestHeaders() {
+		HttpHeaders headers = new HttpHeaders();
+		headers.setAccept(Collections.singletonList(MediaType.APPLICATION_JSON));
+		headers.setContentType(MediaType.APPLICATION_JSON);
+		return headers;
+	}
+
+	private String normalizePath(String path) {
+		return path.startsWith("/") ? path : "/" + path;
+	}
+
+	private <T> ResponseEntity<T> get(String url, Class<T> responseType) {
+		return restTemplate.exchange(url, HttpMethod.GET,
+				new HttpEntity(requestHeaders()), responseType);
+	}
+
+	private <T, R> ResponseEntity<R> post(String url, T requestBody, Class<R> responseType) {
+		HttpEntity<T> requestEntity = new HttpEntity(requestBody, requestHeaders());
+		return restTemplate.exchange(url, HttpMethod.POST, requestEntity, responseType);
+	}
+
+	protected String getActuatorUrl(String deploymentId, String guid) {
+		AppInstanceStatus appInstanceStatus = getDeployedInstance(deploymentId, guid)
+				.orElseThrow(() -> new IllegalStateException(
+						String.format("App with deploymentId %s and guid %s not deployed")));
+		return actuatorUrlForInstance(appInstanceStatus, guid);
+	}
+
+	private Optional<AppInstanceStatus> getDeployedInstance(String deploymentId, String guid) {
+		return appDeployer.status(deploymentId).getInstances().values().stream()
+				.filter(appInstance -> appInstance.getState() == DeploymentState.deployed &&
+						appInstance.getAttributes().get("guid").equals(guid))
+				.findFirst();
```

Comment:
Because we are relying/expecting that the `guid` they pass is in fact unique, should we pick the first one or throw exception if more than one? Could this happen?